src/cwts/networkanalysis/GradientDescentVOSLayoutAlgorithm.java

```diff
@@ -0,0 +1,397 @@
+     */
+    public double getMinStepSize()
+    {
+        return minStepSize;
+    }
+
+    /**
+     * Returns the step size reduction.
+     *
+     * @return Step size reduction
+     */
+    public double getStepSizeReduction()
+    {
+        return stepSizeReduction;
+    }
+
+    /**
+     * Returns the required number of quality value improvements.
+     *
+     * @return Required number of quality value improvements
+     */
+    public int getRequiredNQualityValueImprovements()
+    {
+        return requiredNQualityValueImprovements;
+    }
+
+    /**
+     * Sets the maximum number of iterations.
+     * 
+     * @param maxNIterations Maximum number of iterations
+     */
+    public void setMaxNIterations(int maxNIterations)
+    {
+        this.maxNIterations = maxNIterations;
+    }
+
+    /**
+     * Sets the initial step size.
+     * 
+     * @param initialStepSize Initial step size
+     */
+    public void setInitialStepSize(double initialStepSize)
+    {
+        this.initialStepSize = initialStepSize;
+    }
+
+    /**
+     * Sets the minimum step size.
+     * 
+     * @param minStepSize Minimum step size
+     */
+    public void setMinStepSize(double minStepSize)
+    {
+        this.minStepSize = minStepSize;
+    }
+
+    /**
+     * Sets the step size reduction.
+     * 
+     * @param stepSizeReduction Step size reduction
+     */
+    public void setStepSizeReduction(double stepSizeReduction)
+    {
+        this.stepSizeReduction = stepSizeReduction;
+    }
+
+    /**
+     * Sets the required number of quality value improvements.
+     * 
+     * @param requiredNQualityValueImprovements Required number of quality value
+     *            improvements
+     */
+    public void setRequiredNQualityValueImprovements(int requiredNQualityValueImprovements)
+    {
+        this.requiredNQualityValueImprovements = requiredNQualityValueImprovements;
+    }
+
+    /**
+     * Finds a layout using the gradient descent VOS layout algorithm.
+     *
+     * @param network Network
+     *
+     * @return Layout
+     */
+    public Layout findLayout(Network network)
+    {
+        Layout layout;
+
+        layout = new Layout(network.getNNodes(), random);
+        improveLayout(network, layout);
+        return layout;
+    }
+
+    /**
+     * Improves a layout using the gradient descent VOS layout algorithm.
+     *
+     * @param network Network
+     * @param layout  Layout
+     */
+    public void improveLayout(Network network, Layout layout)
+    {
+        boolean[] visitedNodes;
+        double a, b, distance, distance1, distance2, gradient1, gradient2, gradientLength, qualityValue,
+                oldQualityValue, squaredDistance, stepSize;
+        int i, j, k, l, nQualityValueImprovements;
+        int[] nodeOrder;
+
+        nodeOrder = cwts.util.Arrays.generateRandomPermutation(network.nNodes, random);
+
+        stepSize = initialStepSize;
+        qualityValue = Double.POSITIVE_INFINITY;
```

Comment:
Instead of recalculating the `qualityValue` each time, wouldn't it make more sense to simply use the gradient as an approximation of that?